torch_np/_detail/_reductions.py

```diff
@@ -0,0 +1,164 @@
+""" Implementation of reduction operations, to be wrapped into arrays, dtypes etc
+in the 'public' layer.
+
+Anything here only deals with torch objects, e.g. "dtype" is a torch.dtype instance etc
+"""
+
+import torch
+from . import _util
+from . import _scalar_types
+
+NoValue = None
+
+
+def _atleast_float(dtype, other_dtype):
+    """Return a dtype that is real or complex floating-point.
+
+    For inputs that are boolean or integer dtypes, this returns the default
+    float dtype; inputs that are complex get converted to the default complex
+    dtype; real floating-point dtypes (`float*`) get passed through unchanged
+    """
+    if dtype is None:
+        dtype = other_dtype
+    if not dtype.is_floating_point:
```

Comment:
This not being an `elif` means that if `x` is not a floating point tensor, then it will be casted to a floating point tensor. Is this expected?

Reply:
Exactly. The use is e.g. `tensor.mean()` for `tensor.dtype` being integer. pytorch raises a RuntimeError, but numpy does not. The current version is not 100% numpy-compatible though for edge cases (`np.mean(dtype=int)` etc). Will fix after porting the relevant part of numpy test battery.

Reply:
Libraries allowing implicit casting makes for some incredibly obscure semantics.